src/tiledb/cloud/vcf/ingestion.py

```diff
@@ -0,0 +1,1107 @@
+#!/usr/bin/env python
+
+import configparser
+import enum
+import logging
+import os
+import subprocess
+import threading
+import time
+from collections import defaultdict
+from math import ceil
+from typing import Any, Mapping, Optional, Sequence, Union
+
+import numpy as np
+import tiledbvcf
+
+import tiledb.cloud
+from tiledb.cloud.dag import dag
+
+# --------------------------------------------------------------------
+# Constants
+# --------------------------------------------------------------------
+
+# TODO: automatically run in local mode if the VCFs or the dataset is local
+IS_LOCAL = False
+
+LOG_ARRAY = "log"
+MANIFEST_ARRAY = "manifest"
+
+# TODO: make these configurable
+MANIFEST_BATCH_SIZE = 20
+MANIFEST_WORKERS = 10
+VCF_BATCH_SIZE = 20
+VCF_WORKERS = 10
+VCF_THREADS = 4
+VCF_HEADER_MB = 25  # memory per sample per thread
+
+CONSOLIDATE_CPUS = 4
+CONSOLIDATE_MEMORY_MB = 8192
+
+
+class Contigs(enum.Enum):
+    ALL = enum.auto()
+    CHROMOSOMES = enum.auto()
+    OTHER = enum.auto()
+
+
+# --------------------------------------------------------------------
+# Utilities
+# --------------------------------------------------------------------
+
+
+def read_config(path: str = "~/.aws/credentials"):
+    """Read config values from a file."""
+
+    config = {}
+
+    cp = configparser.ConfigParser()
+    cp.read(os.path.expanduser(path))
+    try:
+        config["vfs.s3.aws_access_key_id"] = cp["default"]["aws_access_key_id"]
+        config["vfs.s3.aws_secret_access_key"] = cp["default"]["aws_secret_access_key"]
+    except KeyError:
+        pass
+
+    return config
```

Comment:
These `utils` can be used in other ingestion functions in the future, but also in some currently implemented e.g. `bioimg`. Could we move it under a `utils` folder for common use?

Reply:
Yes, it's a good idea to share the utility functions and the event logger. I can move them to `tiledb.cloud.utils` or somewhere else, if there's a better place.

Reply:
Reusable functions are now in `tiledb.cloud.utilities` since `utils.py` already existed.

Reply:
> Reusable functions are now in `tiledb.cloud.utilities` since `utils.py` already existed.

`utils` is really for internal consumption only; it could be moved into the `_common` directory.